src-java/swmanager-topology/swmanager-storm-topology/src/main/java/org/openkilda/wfm/topology/switchmanager/fsm/SwitchSyncFsm.java

```diff
@@ -532,46 +540,59 @@ protected void sendRemoveCommands(SwitchSyncState from, SwitchSyncState to,
-            fire(COMMANDS_PROCESSED);
+            fire(NEXT);
             return;
         }
         List<OfCommand> commands = cleanupDependenciesAndBuildCommands(toRemove);
 
         log.info("Remove commands has been sent (switch={}, key={})", switchId, key);
-        carrier.sendOfCommandsToSpeaker(key, commands, OfCommandAction.DELETE, switchId);
+        ofCommandsResponsesCount = carrier.sendOfCommandsToSpeaker(key, commands, OfCommandAction.DELETE, switchId);
     }
 
     protected void sendModifyCommands(SwitchSyncState from, SwitchSyncState to,
                                       SwitchSyncEvent event, Object context) {
         if (toModify.isEmpty()) {
             log.info("No need to process modify commands (switch={}, key={})", switchId, key);
-            fire(COMMANDS_PROCESSED);
+            fire(NEXT);
             return;
         }
         List<OfCommand> commands = cleanupDependenciesAndBuildCommands(toModify);
 
         log.info("Modify commands has been sent (switch={}, key={})", switchId, key);
-        carrier.sendOfCommandsToSpeaker(key, commands, OfCommandAction.MODIFY, switchId);
+        ofCommandsResponsesCount = carrier.sendOfCommandsToSpeaker(key, commands, OfCommandAction.MODIFY, switchId);
     }
 
     protected void sendInstallCommands(SwitchSyncState from, SwitchSyncState to,
                                        SwitchSyncEvent event, Object context) {
         if (toInstall.isEmpty()) {
             log.info("No need to process install commands (switch={}, key={})", switchId, key);
-            fire(COMMANDS_PROCESSED);
+            fire(NEXT);
             return;
         }
         List<OfCommand> commands = cleanupDependenciesAndBuildCommands(toInstall);
 
         log.info("Install commands has been sent (switch={}, key={})", switchId, key);
-        carrier.sendOfCommandsToSpeaker(key, commands, OfCommandAction.INSTALL, switchId);
+        ofCommandsResponsesCount = carrier.sendOfCommandsToSpeaker(key, commands, OfCommandAction.INSTALL, switchId);
+    }
+
+    protected void ofCommandsProcessed(SwitchSyncState from, SwitchSyncState to,
+                                       SwitchSyncEvent event, Object context) {
+        log.info("OF commands processed (switch={}, key={})", switchId, key);
+        ofCommandsResponsesCount--;
+        if (ofCommandsResponsesCount <= 0) {
+            fire(NEXT);
+        }
     }
 
     private List<OfCommand> cleanupDependenciesAndBuildCommands(List<SpeakerData> source) {
         Set<UUID> ids = source.stream()
                 .map(SpeakerData::getUuid)
                 .collect(Collectors.toSet());
         source.forEach(speakerData -> speakerData.getDependsOn().retainAll(ids));
+
+        // As we send commands by several batches we need to sort the list to do not send batch with command A before
+        // batch with command B if A depends on B.
+        source = RuleManagerHelper.sortCommandsByDependencies(source);
```

Comment:
If A depends on B and A will be in first batch and B in second we will probably get some "Dependency not found error" for second batch. I guess we need to cleanup dependencies after batch split.

Reply:
I guess you missed "not" in comment. 
```
As we send commands by several batches we need to sort the list to do NOT send batch with command A before
        // batch with command B if A depends on B.
```
If A depends on B we must send bath with B first

Reply:
the problem was fixed